Flow test fails with samples between 3000 and 100000

I am just opening this issue to document and track the progress that i've done so far with bigger samples. The initial problem is set to 3000 and I set an arbitrary limit to this issue at 100 000.

Here are my results for 10 000:
```
	Maximum of the shovels is: 10116

		TEST RESULTS

                 | dd.weather routing |
test  1 success: sr_shovel	 (10116) t_dd1 should have the same number of items as t_dd2	 (10116)
test  2 success: sr_winnow	 (20232) should have the same of the number of items of shovels	 (20232)
test  3 success: sr_sarra	 (10088) should have the same number of items as winnows'post	 (10101)
test  4 success: sr_subscribe	 (10088) should have the same number of items as sarra		 (10088)
                 | watch      routing |
test  5 FAILURE: sr_watch		 (27117) should be 4 times subscribe t_f30		  (10088)
test  6 success: sr_sender		 (26316) should have the same number of items as sr_watch  (27117)
test  7 success: sr_subscribe u_sftp_f60 (26329) should have the same number of items as sr_sender (26316)
test  8 success: sr_subscribe cp_f61	 (26329) should have the same number of items as sr_sender (26316)
                 | poll       routing |
test  9 success: sr_poll test1_f62	 (15529) should have half the same number of items of sr_sender	 (26316)
test 10 success: sr_subscribe q_f71	 (15541) should have the same number of items as sr_poll test1_f62 (15529)
                 | flow_post  routing |
test 11 success: sr_post test2_f61	 (11467) should have half the same number of items of sr_sender 	 (26316)
test 12 success: sr_subscribe ftp_f70	 (11467) should have the same number of items as sr_post test2_f61 (11467)
test 13 success: sr_post test2_f61	 (11467) should have about the same number of items as shim_f63	 (11464)
                 | py infos   routing |
test 14 FAILURE: sr_shovel pclean_f90 (5683) should have the same number of watched items winnows' post	 (10101)
test 15 FAILURE: sr_shovel pclean_f92 (5046) should have the same number of removed items winnows' post	 (10101)
test 16 success: 0 messages received that we don't know what happened.
test 17 success: count of truncated headers (10088) and subscribed messages (10088) should have about the same number of items
                 | C          routing |
test 18 success: cpump both pelles (c shovel) should receive about the same number of messages (42892) (42892)
test 19 success: cdnld_f21 subscribe downloaded (3081) the same number of files that was published by both van_14 and van_15 (3081)
test 20 FAILURE: veille_f34 should post twice as many files (3884) as subscribe cdnld_f21 downloaded (3081)
test 21 success: veille_f34 should post twice as many files (3884) as subscribe cfile_f44 downloaded (1655)
test 22 FAILURE: Overall 17 of 21 passed (sample size: 10101) !
```

... and here is the some error encountered:
- From sr_cpost
```
sr_cpost_veille_f34_01.log [ERROR] info: loop detected, skipping: /local/ben/sarra_devdocroot/cfr/bulletins/alphanumeric/20190410/SR/KWAL
sr_cpost_veille_f34_01.log [ERROR] failed to add_watch: /local/ben/sarra_devdocroot/cfr/bulletins/alphanumeric/20190410/SR
sr_cpost_veille_f34_01.log [ERROR] failed to add_watch: /local/ben/sarra_devdocroot/cfr/bulletins/alphanumeric/20190410
sr_cpost_veille_f34_01.log [ERROR] failed to add_watch: /local/ben/sarra_devdocroot/cfr/bulletins/alphanumeric
sr_cpost_veille_f34_01.log [ERROR] failed to add_watch: /local/ben/sarra_devdocroot/cfr/bulletins
```
- From sr_sender
```
sr_sender_tsource2send_f50_01.log [ERROR] sender/__do_send__: could not send
sr_sender_tsource2send_f50_01.log [ERROR] Delivery failed /local/ben/sarra_devdocroot/sent_by_tsource2send/20190410/WXO-DD/bulletins/alphanumeric/20190410/SA/KWBC
```
- From sr_shovel_pclean:
```
5 sr_shovel_pclean_f90_01.log [ERROR] sr_shovel/run going badly, so sleeping for 0.##
```
- From sr_subscribe_cclean:
```
sr_subscribe_cclean_f91_01.log [ERROR] sr_consumer/consume malformed message {'frame_method': (60, 71), 'frame_args': b'\x00<\x00G\x00\x00\x00\x00\x00\x00\x03%\x00\x08xcpublicRv02.post.local.ben.sarra_devdocroot.cfr.bulletins.alphanumeric.20190410.SI.CWVR.21\x00\x00\x02x', 'properties': {'content_type': 'text/plain', 'application_headers': {'from_cluster': 'localhost', 'sum': 'R,62b9f8c512899cccad9afb9f5af8ae591f36e2b0588fd02510735eb543fcd5167f5058f468ec3cfb56cc4cfbbd43bda37f3dbf2496e5895139d15a70367ab9f0', 'to_clusters': 'localhost'}, 'delivery_mode': 2}, '_pending_chunks': [], 'body_received': 110, 'body_size': 110, 'ready': True, 'delivery_info': {'delivery_tag': 805, 'redelivered': False, 'exchange': 'xcpublic', 'routing_key': 'v02.post.local.ben.sarra_devdocroot.cfr.bulletins.alphanumeric.20190410.SI.CWVR.21', 'message_count': 632}, 'body': b'20190410214003.053809449 file:/ /local/ben/sarra_devdocroot/cfr/bulletins/alphanumeric/20190410/SI/CWVR/21/\xcc \n', 'channel': <amqp.channel.Channel object at 0x7fcb3714c6d8>, 'isRetry': False}
sr_subscribe_cclean_f91_01.log [ERROR] sr_subscribe/run going badly, so sleeping for 0.01
```
- Also got that from various sr_subscribe:
```
sr_subscribe_cp_f61_03.log [ERROR] did not move /local/ben/sarra_devdocroot/downloaded_by_sub_cp/20190410/WXO-DD/bulletins/alphanumeric/20190410/SR/KWAL/21/SRMN20_KWAL_102134___42260 to /local/ben/sarra_
devdocroot/downloaded_by_sub_cp/20190410/WXO-DD/bulletins/alphanumeric/20190410/SR/KWAL/21/SRMN20_KWAL_102134___42260.moved (rename)
sr_subscribe_cp_f61_03.log [ERROR] sr_subscribe/doit_download: could not remove /local/ben/sarra_devdocroot/downloaded_by_sub_cp/20190410/WXO-DD/bulletins/alphanumeric/20190410/SP.
sr_subscribe_cp_f61_05.log [ERROR] did not move /local/ben/sarra_devdocroot/downloaded_by_sub_cp/20190410/WXO-DD/bulletins/alphanumeric/20190410/SX/KWAL/21/SXAK50_KWAL_102124___20998 to /local/ben/sarra_
devdocroot/downloaded_by_sub_cp/20190410/WXO-DD/bulletins/alphanumeric/20190410/SX/KWAL/21/SXAK50_KWAL_102124___20998.moved (rename)
sr_subscribe_cp_f61_05.log [ERROR] Download failed sftp://ben@localhost//local/ben/sarra_devdocroot/sent_by_tsource2send/20190410/WXO-DD/bulletins/alphanumeric/20190410/SP/CYBB/21/SPCN62_CYBB_102138___30
85961
sr_subscribe_q_f71_04.log [ERROR] Download failed sftp://ben@localhost///local/ben/sarra_devdocroot/sent_by_tsource2send/20190410/WXO-DD/citypage_weather/xml/AB/s0000781_f.xml.slink
sr_subscribe_q_f71_05.log [ERROR] Download failed sftp://ben@localhost///local/ben/sarra_devdocroot/sent_by_tsource2send/20190410/WXO-DD/citypage_weather/xml/NT/s0000443_e.xml.slink
sr_subscribe_q_f71_05.log [ERROR] Download failed sftp://ben@localhost///local/ben/sarra_devdocroot/sent_by_tsource2send/20190410/WXO-DD/citypage_weather/xml/AB/s0000794_f.xml.slink
sr_subscribe_t_f30_01.log [ERROR] util/writelocal mismatched file length writing s0000086_e.xml.tmp. Message said to expect 28600 bytes.  Got 28323 bytes.
```

--- a/sarra/sr_post.py
+++ b/sarra/sr_post.py
@@ -1165,25 +1165,9 @@ class sr_post(sr_instances):
            if sys.platform == 'win32':
                src = src.replace('\\','/')
 
-        # walk src directory, this walk is depth first... there could be a lot of time
-        # between *listdir* run, and when a file is visited, if there are subdirectories before you get there.
-        # hence the existence check after listdir (crashed in flow_tests of > 20,000)
-        try:
-            dir_content = os.listdir(src)
-        except FileNotFoundError as err:
-            self.logger.error("could not list dir({}): {}".format(src, err))
-            self.logger.debug("Exception details:", exc_info=True)
-            dir_content = []
-
-        for x in dir_content:
-            path = src + '/' + x
-            if os.path.isdir(path):
-               self.walk(path)
-               continue
-
-            # add path created
-            if os.path.exists(path):
-                self.post1file(path,os.stat(path))
+        for path in glob.glob(os.path.join(src, '**', '*'), recursive=True):
+            if os.path.isfile(path):
+                self.post1file(path, os.stat(path))
 
     # =============
     # original walk_priming

--- a/sarra/sr_post.py
+++ b/sarra/sr_post.py
@@ -1026,8 +1026,13 @@ class sr_post(sr_instances):
         if not os.path.exists(src) : return done
 
         # file : must be old enough
+        try:
+            lstat = os.stat(src)
+        except FileNotFoundError as err:
+            self.logger.error("could not stat file ({}): {}".format(src, err))
+            self.logger.debug("Exception details:", exc_info=True)
+            return done
 
-        lstat = os.stat(src)
         if self.path_inflight(src,lstat): return later
 
         # post it
@@ -1163,7 +1168,14 @@ class sr_post(sr_instances):
         # walk src directory, this walk is depth first... there could be a lot of time
         # between *listdir* run, and when a file is visited, if there are subdirectories before you get there.
         # hence the existence check after listdir (crashed in flow_tests of > 20,000)
-        for x in os.listdir(src):
+        try:
+            dir_content = os.listdir(src)
+        except FileNotFoundError as err:
+            self.logger.error("could not list dir({}): {}".format(src, err))
+            self.logger.debug("Exception details:", exc_info=True)
+            dir_content = []
+
+        for x in dir_content:
             path = src + '/' + x
             if os.path.isdir(path):
                self.walk(path)

--- a/sarra/sr_post.py
+++ b/sarra/sr_post.py
@@ -538,7 +538,12 @@ class sr_post(sr_instances):
            else:
               binary = (self.inline_encoding == 'text' )
 
-           f = open(path,'rb')
+           try:
+               f = open(path,'rb')
+           except FileNotFoundError as err:
+               self.logger.error("could not open ({}): {}".format(path, err))
+               self.logger.debut("Exception details:", exc_info=True)
+               return False
            d = f.read()
            f.close()
 
@@ -1189,10 +1194,13 @@ class sr_post(sr_instances):
             d=p
 
         try :
-                 fs = os.stat(d)
-                 dir_dev_id = '%s,%s' % ( fs.st_dev, fs.st_ino )
-                 if dir_dev_id in self.inl: return True
-        except:  self.logger.warning("could not stat %s" % d)
+            fs = os.stat(d)
+            dir_dev_id = '%s,%s' % ( fs.st_dev, fs.st_ino )
+            if dir_dev_id in self.inl:
+                return True
+        except FileNotFoundError as err:
+            self.logger.warning("could not stat file ({}): {}".format(d, err))
+            self.logger.debug("Exception details:", exc_info=True)
 
         if os.access( d , os.R_OK|os.X_OK ): 
            try:

--- a/sarra/sr_post.py
+++ b/sarra/sr_post.py
@@ -1158,14 +1158,11 @@ class sr_post(sr_instances):
     def walk(self, src ):
         self.logger.debug("walk %s" % src )
 
-        # how to proceed with symlink
-
-        if os.path.islink(src) and self.realpath_post :
-           src = os.path.realpath(src)
-           if sys.platform == 'win32':
-               src = src.replace('\\','/')
-
-        for path in glob.glob(os.path.join(src, '**', '*'), recursive=True):
+        for path in glob.iglob(os.path.join(src, '**', '*'), recursive=True):
+            if self.realpath_post and os.path.isfile(path):
+                path = os.path.realpath(path)
+                if sys.platform == 'win32':
+                    path = path.replace('\\', '/')
             if os.path.isfile(path):
                 self.post1file(path, os.stat(path))
 

--- a/test/flow_limit.sh
+++ b/test/flow_limit.sh
@@ -65,22 +65,22 @@ printf  "\nSufficient!\n"
 if [ ! "$SARRA_LIB" ]; then
    if [ "`sr_shovel t_dd1_f00 status |& tail -1 | awk ' { print $8 } '`" != 'stopped' ]; then 
        echo "Stopping shovels and waiting..."
-       sr_shovel stop t_dd2_f00 &
+       sr_shovel stop t_dd2_f00
        sr_shovel stop t_dd1_f00 
    fi
 else 
    if [ "`$SARRA_LIB/sr_shovel.py t_dd1_f00 status |& tail -1 | awk ' { print $8 } '`" != 'stopped' ]; then
        echo "Stopping shovels and waiting..."
-       "$SARRA_LIB"/sr_shovel.py stop t_dd2_f00 &
+       "$SARRA_LIB"/sr_shovel.py stop t_dd2_f00
        "$SARRA_LIB"/sr_shovel.py stop t_dd1_f00
    fi
 fi
 
 if [ "$SARRAC_LIB" ]; then
-   "$SARRAC_LIB"/sr_cpump stop pelle_dd1_f04 &
+   "$SARRAC_LIB"/sr_cpump stop pelle_dd1_f04
    "$SARRAC_LIB"/sr_cpump stop pelle_dd2_f05
 elif [ "${C_ALSO}" ]; then
-   sr_cpump stop pelle_dd1_f04 &
+   sr_cpump stop pelle_dd1_f04
    sr_cpump stop pelle_dd2_f05
 fi
 

